File: connectors/connector-msp/src/main/scala/com/taskadapter/connector/msp/CustomFieldConverter.scala
package com.taskadapter.connector.msp

/**
  * A custom field value can be anything, e.g. Seq[String] for "label" type loaded from JIRA.
  * That field would have a list of strings. This class converts those values to a single string
  * that can be used in Microsoft Project strings-based fields.
  */
object CustomFieldConverter {

  /**
    * @return empty string if given value is null, otherwise converted value. never returns null
    */
  def getValueAsString(value: Any): String = {
    if (value == null) {
      return ""
    }
    if (value.isInstanceOf[Seq[String]]) {
      val seq = value.asInstanceOf[Seq[String]]
      seq.mkString(" ")
    } else if (value.isInstanceOf[scala.collection.immutable.Seq[String]]) {
      val seq = value.asInstanceOf[scala.collection.immutable.Seq[String]]
      seq.mkString(" ")
    } else {
      value.toString
    }
  }
}

File: connectors/connector-msp/src/main/java/com/taskadapter/connector/msp/CustomFieldConverter.java
package com.taskadapter.connector.msp;

import java.util.List;

public class CustomFieldConverter {

    /**
     * @return empty string if given value is null, otherwise converted value. never returns null
     */
    public static String getValueAsString(Object value) {
        if (value == null) {
            return "";
        }
        if (value instanceof List<?>) {
            List<?> list = (List<?>) value;
            if (list.isEmpty()) {
                return "";
            }
            var first = list.get(0);
            if (first instanceof String) {
                return String.join(" ", (List<String>) list);
            }
            return "unknown type, only list of strings are supported";
        }
        return value.toString();
    }
}
